### PR TITLE
packer: relax constraints on local sources

### PR DESCRIPTION
**Pull request by lbajolet-hashicorp**
The source parsing logic was heavily directed towards Github compatible source URIs, however if we want to support more cases, we need to make sure we are able to specify those URIs, and to load plugins installed from those sources.

Right now, since the getters available are only github.com, we will not support remotely instlling plugins from sources other than github.com, with the same set of constraints as before. However, we do support now installing from a local plugin binary to any kind of source, and we support loading them, including if a template wants this plugin installed locally with version constraints.

Related to: #12863

### CODE REVIEW

**Comment by sylviamoss** [comment]
> However, we do support now installing from a local plugin binary to any kind of source, and we support loading them, including if a template wants this template installed locally with version constraints.

I need help understanding this part, can you show me an example?

File: hcl2template/addrs/plugin.go [hunk 1]
```diff
@@ -13,21 +16,58 @@ import (
...
+	}
+	retParts = append(retParts, part)
+
+	// The algorithm above registers the path in reversed order
+	//
+	// The function is supposed to return them in the host->...->name
+	// order however, so we have to reverse the slice.
+	// There isn't (afaik) a standard function for that in Go, so we write
+	// it manually.
+	reversedParts := make([]string, len(retParts))
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:51]: Seeing as Packer is using Go 1.21. The slices pkg has a Reverse function you can use here.

File: hcl2template/addrs/plugin.go [hunk 2]
```diff
@@ -13,21 +16,58 @@ import (
...
+	// misbehave.
+	if p.Source == "" {
+		log.Printf("[ERROR] - Plugin.Parts: the source was empty, this should not happen in normal usage, and is likely a Packer bug.")
+		return retParts
+	}
+
+	dir, part := path.Split(strings.TrimRight(p.Source, "/"))
+	for dir != "" {
+		retParts = append(retParts, part)
+		dir, part = path.Split(strings.TrimRight(dir, "/"))
```
[the remark this answers is not in the document]
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:41]: So, `dir` is never empty in that loop, since we are purposefully looping until it becomes empty.
That happens when you split the path, as when you reach the top dir, it becomes empty.

Ex:

```
p.Source = "github.com/hashicorp/amazon"
# Loop 1
dir = "github.com/hashicorp"
part = "amazon"
# Loop 2
dir = "github.com"
part = "hashicorp"
# Loop 3
dir = ""
part = "github.com"
# Out of loop

retParts = ["amazon", "hashicorp", "github.com"]
```

**Comment by lbajolet-hashicorp** [comment]
> > However, we do support now installing from a local plugin binary to any kind of source, and we support loading them, including if a template wants this template installed locally with version constraints.
> 
> I need help understanding this part, can you show me an example?

Duplicate `template` in the message, it should read "including if a template wants this plugin installed locally with version constraints".

What that means essentially (and the problem we're trying to solve here), is if someone had a plugin that cannot be sourced through `github.com`, they could still install it manually as `packer-plugin-<name>`. Since we're removing this capacity with 1.11.0, that means they'd need to make their source something like `github.com/org/name`, which may not necessarily fit the expected source location of the plugin binaries.
Now they cannot be installed still, as we don't expose any getter other than github, but at least relaxing the constraints on the source means that they can have whatever makes sense for a source URI (so anything that is essentially a hostname and a resource path, even minimal), and they can install them locally, and use `required_plugins` blocks to manage which version of a plugin is required by their template.

File: hcl2template/addrs/plugin.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sylviamoss) [reply hcl2template/addrs/plugin.go:42]: Ohh, I didn't see it was a `for`, my bad!!

**Review by sylviamoss — APPROVED** [review]
👍🏼 LGTM

File: hcl2template/addrs/plugin.go [hunk 3]
```diff
@@ -13,21 +17,53 @@ import (
...
+	}
+	retParts = append(retParts, part)
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:44]: I don't understand why we are doing this. Can you explain.

File: hcl2template/addrs/plugin.go [hunk 4]
```diff
@@ -13,21 +17,53 @@ import (
...
-	Hostname  string
-	Namespace string
-	Type      string
+	Source string
 }
 
-func (p Plugin) RealRelativePath() string {
-	return p.Namespace + "/packer-plugin-" + p.Type
+// Parts returns the list of components of the source URL, starting with the
+// host, and ending with the name of the plugin (non-simplified).
```
[the remark this answers is not in the document]
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:24]: Oh that might be out-of-scope given that I removed the `packer-plugin-<name>` support from the URIs after we discussed it out-of-band.

The `non-simplified` part here meant that the name of the plugin might be in the `packer-plugin-amazon` form given that I had removed the check during installation, but since it's been reverted, the `name` should always be `packer-plugin-` free now.

File: hcl2template/addrs/plugin.go [hunk 5]
```diff
@@ -98,67 +134,63 @@ func IsPluginPartNormalized(str string) (bool, error) {
...
+	str = strings.Trim(str, "/")
+
+	url, err := url.Parse(str)
+	if err != nil {
+		return nil, diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,
-			Summary:  "Invalid plugin source string",
-			Detail:   `The "source" attribute must be in the format "hostname/namespace/name"`,
+			Summary:  "Invalid source URL",
+			Detail:   "A source must be a valid URL",
```
[the remark this answers is not in the document]
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:150]: Yep without a scheme as I believe we do check it? Even without scheme this may be a valid URL, but I agree the error message can be improved, I'll see what I can reroll there

File: hcl2template/addrs/plugin.go (hunk 3 above)
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:44]: The last `append` you mean? It's because the last `split` above returns an empty `dir` and a non-empty `part`, so we leave the loop, therefore that last `part` is not added to the slice, hence why we call it once more after that.

To illustrate:

```
p.Source = "github.com/hashicorp/amazon"
# Loop 1
dir = "github.com/hashicorp"
part = "amazon"
retParts = ["amazon"]
# Loop 2
dir = "github.com"
part = "hashicorp"
retParts = ["amazon", "hashicorp"]
# Loop 3
dir = ""
part = "github.com"
retParts = ["amazon", "hashicorp"]
# Out of loop
retParts = ["amazon", "hashicorp", "github.com"]]
```

Comment (nywilken) [reply hcl2template/addrs/plugin.go:44]: Sorry I should've been more specific with my questions. I was trying to figure out why you split the string this way. 

Looking at it again I suspect you want to split after each path separator '/' but since you don't know the count you keep calling pat.Split until you have all parts. Is that correct? 

If so, we can achieve this differently without needing the final append or slice.Reverse. 

```go
	source := strings.TrimRight(p.Source, "/")
	parts := strings.FieldsFunc(source, func(c rune) bool {
		return c == '/'
	})
```

File: hcl2template/addrs/plugin.go [hunk 6]
```diff
@@ -13,21 +17,53 @@ import (
...
+		retParts = append(retParts, part)
+		dir, part = path.Split(strings.TrimRight(dir, "/"))
+	}
+	retParts = append(retParts, part)
+
+	// The algorithm above registers the path in reversed order
+	//
+	// The function is supposed to return them in the host->...->name
+	// order however, so we have to reverse the slice.
+	slices.Reverse(retParts)
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:50]: By using the strings.FieldFunc to split we can get ride of this reverse.

File: hcl2template/addrs/plugin.go [hunk 7]
```diff
@@ -13,21 +17,53 @@ import (
...
-func (p Plugin) Parts() []string {
-	return []string{p.Hostname, p.Namespace, p.Type}
+// Name returns the raw name of the plugin from its source
+//
+// Exemples:
+//   - "github.com/hashicorp/amazon" -> "amazon"
+//   - "github.com/hashicorp/packer-plugin-amazon" -> "amazon"
+func (p Plugin) Name() string {
+	parts := p.Parts()
+	return strings.ReplaceAll(parts[len(parts)-1], "packer-plugin-", "")
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:62]: I believe this still supports plugin Source addresses in the form of "github.com/hashicorp/packer-plugin-docker" right. Lets break this out into separate PR so we can validate it doesn't breaking anything given packer init original implementation.

File: hcl2template/addrs/plugin.go [hunk 8]
```diff
@@ -13,21 +17,53 @@ import (
...
+// This will correspond more or less to the filesystem hierarchy where
+// the plugin is installed.
+func (p Plugin) Parts() []string {
+	var retParts []string
+
+	// This should not happen, we're supposed to check that the source isn't
+	// empty when creating the plugin, but let's be careful here and not
+	// misbehave.
+	if p.Source == "" {
+		log.Printf("[ERROR] - Plugin.Parts: the source was empty, this should not happen in normal usage, and is likely a Packer bug.")
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:35]: I see why logging is nice for debugging but it seem out of place given the use of this Method, at least to me. 

What about documenting the behavior in the comments.
```go
// Parts splits the plugin Source by URL path separator separating 
// it into hostname, namespace directories, and plugin name. If there 
// is no slash a slice containing the original is returned. On empty plugin Sources
// an empty slice is returned. It is up to the caller to check the length of the resulting slice.
func (p Plugin) Parts() []string {
```

File: hcl2template/addrs/plugin.go [hunk 9]
```diff
@@ -98,67 +134,63 @@ func IsPluginPartNormalized(str string) (bool, error) {
...
-		Hostname:  "",
-		Namespace: "",
-	}
 	var diags hcl.Diagnostics
 
-	// split the source string into individual components
-	parts := strings.Split(str, "/")
-	if len(parts) != 3 {
-		diags = diags.Append(&hcl.Diagnostic{
+	// Trim both leading and trailing slashes for convenience
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:142]: I suggest adding a call out on why we Trim,as apposed to calling out we trim for convenience. Looking at the parts function we want to make sure that the plugin source is in a valid addrs format and not a root file path. Is that correct?

File: hcl2template/addrs/plugin.go [hunk 10]
```diff
@@ -98,67 +134,63 @@ func IsPluginPartNormalized(str string) (bool, error) {
...
-			Detail:   fmt.Sprintf(`Invalid plugin type %q in source %q: %s"`, givenName, str, err),
+			Summary:  "Invalid source URL",
+			Detail:   "A source must not contain a query or a fragment",
+		})
+	}
+
+	str = url.String()
+
+	slashCount := strings.Count(str, "/")
+	if slashCount == 0 {
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:173]: I feel like this check should happen higher up but as I understand you are validating the url first then checking of paths, thats why it is here. You can achieve the same results using strings.Contains if slashCount is not used anywhere else.

File: hcl2template/addrs/plugin.go (hunk 3 above)
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:44]: Strings may do the trick, unless we have an escaped `/` in the middle of the path (though I guess in valid URLs they should be URL-encoded?).
I went with `path.Split` since it's URL-centric, and I trust it will do the splitting according to URL-rules (all of which I don't know).

But you are right: conceptually I went with this approach as we don't know how many parts there are in that URL, and I didn't find a convenience method to explode the URL in its parts.

File: hcl2template/addrs/plugin.go (hunk 7 above)
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:62]: So theoretically the structure allows it, we could get rid of that `ReplaceAll` though.

The check is done at creation-time, in the `ParsePluginSourceString` function, there we check if the name contains `packer-` or `packer-plugin-`, and we error.
It's not bulletproof as the structure can still be created from scratch, so that protection could be circumvented (this was already the case though to be clear, it's not something we are introducing right now)

File: hcl2template/addrs/plugin.go (hunk 8 above)
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:35]: We can document the usage in the function itself yes, but that doesn't negate the fact that creating a `Plugin` instance with an empty string is a misuse of the structure, which we should probably be able to spot one way or another.
I'm not a fan of having `Parts` error/panic as if we have a bug this could be disruptive to users, but I think it's still valuable to at least warn through verbose logs that something went wrong.

Though I guess that implies users will need to look at the verbose logs to diagnose the problem (if perceived), so maybe this is unnecessary, unsure about this one.

File: hcl2template/addrs/plugin.go (hunk 9 above)
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:142]: Yeah good call out. I used `Trim` here to drop the leading/trailing slashes so the `url.Parse` returns something expected in our case. Technically having a URL that starts with `//` or that ends with `/` is valid, but throws off our parsing, so I defensively remove them here to not have to deal with special cases like those later on.

Though I presume we could alternatively explicitely reject URIs formed like this, this might be clearer for users.

File: hcl2template/addrs/plugin.go (hunk 3 above)
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:44]: Note: after talking out-of-band and looking into the actual implementation, `path.Split` doesn't do much more than a `strings.Split` so we can go with this implementation instead.

File: acctest/plugin/plugin_acc_test.go [hunk 11]
```diff
@@ -69,27 +67,31 @@ func TestAccInitAndBuildBasicAmazonEbs(t *testing.T) {
...
 
-func cleanupPluginInstallation(plugin addrs.Plugin) error {
+func pluginDirectory(plugin addrs.Plugin) (string, error) {
 	home, err := homedir.Dir()
 	if err != nil {
-		return err
+		return "", err
 	}
-	pluginPath := filepath.Join(home,
+	if xdgConfigHome := os.Getenv("XDG_CONFIG_HOME"); xdgConfigHome != "" {
```
[the remark this answers is not in the document]
Comment (lbajolet-hashicorp) [reply acctest/plugin/plugin_acc_test.go:75]: Yep, that's essentially what `pathing.configDir` does (for UNIX systems at least).
That code I changed since it did not support plugins with a depth !=3, but truth be told rather than trying to figure ourselves what the expected directory is supposed to be here (with OS-specific code), we can lean on `packer.PluginFolder`

File: acctest/plugin/component_acc_test.go [hunk 12]
```diff
@@ -22,10 +23,9 @@ var basicAmazonAmiDatasourceHCL2Template string
 
 func TestAccInitAndBuildBasicAmazonAmiDatasource(t *testing.T) {
 	plugin := addrs.Plugin{
-		Hostname:  "github.com",
-		Namespace: "hashicorp",
-		Type:      "amazon",
+		Source: "github.com/hashicorp/amazon",
 	}
+	log.Printf("Changed")
```
Comment (nywilken) [inline acctest/plugin/component_acc_test.go:28]: Oops - something left over from debug session.

File: acctest/plugin/plugin_acc_test.go [hunk 13]
```diff
@@ -69,27 +67,24 @@ func TestAccInitAndBuildBasicAmazonEbs(t *testing.T) {
...
-			plugin.Hostname,
-			plugin.Namespace,
-			plugin.Type)
-	}
+	pluginParts := []string{pluginDir}
+	pluginParts = append(pluginParts, plugin.Parts()...)
+	pluginPath := filepath.Join(pluginParts...)
 
+	return pluginPath, nil
+}
```
Comment (nywilken) [inline acctest/plugin/plugin_acc_test.go:81]: nit: We can simplify this a little
```suggestion
	return filepath.Join(pluginParts...), nil
}
```

File: hcl2template/addrs/plugin.go [hunk 14]
```diff
@@ -98,67 +110,71 @@ func IsPluginPartNormalized(str string) (bool, error) {
...
-	parts := strings.Split(str, "/")
-	if len(parts) != 3 {
-		diags = diags.Append(&hcl.Diagnostic{
+	if strings.HasPrefix(str, "/") {
+		return nil, diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,
-			Summary:  "Invalid plugin source string",
-			Detail:   `The "source" attribute must be in the format "hostname/namespace/name"`,
+			Summary:  "Invalid source URL",
+			Detail:   `The source URL must not start with a '/' character. Ex: "github.com/hashicorp/amazon"`,
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:122]: To not give the impression that one cloud is favored over the other lets use the generic happycloud
```suggestion
			Detail:   `The source URL must not start with a '/' character. Ex: "github.com/hashicorp/happycloud"`,
```

File: hcl2template/addrs/plugin.go [hunk 15]
```diff
@@ -98,67 +110,71 @@ func IsPluginPartNormalized(str string) (bool, error) {
...
+
+	url, err := url.Parse(str)
 	if err != nil {
-		diags = diags.Append(&hcl.Diagnostic{
+		return nil, diags.Append(&hcl.Diagnostic{
 			Severity: hcl.DiagError,
-			Summary:  "Invalid plugin type",
-			Detail:   fmt.Sprintf(`Invalid plugin type %q in source %q: %s"`, givenName, str, err),
+			Summary:  "Invalid source URL",
+			Detail:   fmt.Sprintf(`A source must be a valid URL: %s. Ex: "github.com/hashicorp/amazon"`, err),
```
Comment (nywilken) [inline hcl2template/addrs/plugin.go:147]: Anyone can argue that a valid URL does contain x,y,z. Lets just tell the user what is expected. 
```suggestion
			Detail:   fmt.Sprintf(`A source must be in the expected format: %s. Ex: "github.com/hashicorp/amazon"`, err),
```

File: hcl2template/addrs/plugin_test.go [hunk 16]
```diff
@@ -6,29 +6,109 @@ package addrs
...
 
-func TestParsePluginSourceString(t *testing.T) {
-	type args struct {
-		str string
-	}
+func TestPluginParseSourceString(t *testing.T) {
 	tests := []struct {
-		args      args
+		name      string
+		str       string
```
Comment (nywilken) [inline hcl2template/addrs/plugin_test.go:16]: Any reason why you just didn't call this source?

File: packer/plugin-getter/github/getter.go [hunk 17]
```diff
@@ -156,10 +158,40 @@ func (t *HostSpecificTokenAuthTransport) base() http.RoundTripper {
 	return http.DefaultTransport
 }
 
+type GithubPlugin struct {
```
Comment (nywilken) [inline packer/plugin-getter/github/getter.go:161]: nit: Lets call this what it is a plugin URI for GitHub. 
```suggestion
type GitHubPluginURI struct {
```

File: packer/plugin-getter/github/getter.go [hunk 18]
```diff
@@ -156,10 +158,40 @@ func (t *HostSpecificTokenAuthTransport) base() http.RoundTripper {
...
+	Hostname  string
+	Namespace string
+	Type      string
+}
+
+func PluginToGithubURI(source *addrs.Plugin) (*GithubPlugin, error) {
```
Comment (nywilken) [inline packer/plugin-getter/github/getter.go:167]:
```suggestion
func NewGitHubPluginURI(source *addrs.Plugin) (GithubPlugin, error) {
```

I think you can also just take in three arguments here (hostname, namespace, reponame string) and drop the validation for three parts and the passing of an addrs.Plugin. Or just take in a string and split on to validate three parts.

**Review by nywilken — APPROVED** [review]
The re-roll looks good. I left a few more nits and suggestions. Feel free to merge when ready.

File: hcl2template/addrs/plugin.go (hunk 15 above)
Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:147]: I'm not certain the error message is clearer here tbh, we URL parse the source and return the error from the go library, is "must be in the expected format" clearer than "must be a valid URL"?

Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:147]: To be clear, I agree with you that "valid URL" is not the only constraint here and just receiving "must be a valid URL" is not too accurate: I'm not sure about the suggested reword, not that it doesn't need one.

Comment (lbajolet-hashicorp) [reply hcl2template/addrs/plugin.go:147]: Alternatively, what if we returned a list of errors rather than stop at the first one? This way a user will have all the errors that they need to act on. Some need prerequisites to work (like schema detection or query/fragment need a parsed URL), but this may be helpful still?

File: packer/plugin-getter/github/getter.go (hunk 17 above)
Comment (lbajolet-hashicorp) [reply packer/plugin-getter/github/getter.go:161]: Hmmmm, since the general structure is called `Plugin`, should we rename both to be `URI`? This would make more sense probably.
I'm hesitant in changing this one without changing the other, if only for consistency's sake.

File: packer/plugin-getter/github/getter.go (hunk 18 above)
Comment (lbajolet-hashicorp) [reply packer/plugin-getter/github/getter.go:167]: We could take three arguments, but in the end the splitting into three parts will need to be done at callsite instead of in that function, I'm not sure this is better?
Taking the source string works as well, but since the `addrs.Plugin` type offers a `Parts()` function that does this well for us, making this work on any string will essentially involve re-doing that code locally, so I'm not certain of this alternative. (Besides maybe testing? Although it's easy to make a plugin with a source string in that context).
